sympy/polys/domains/quaternion.py

```diff
@@ -0,0 +1,524 @@
+
+        b = Quaternion.__copysign(b, M[2, 1] - M[1, 2])
+        c = Quaternion.__copysign(c, M[0, 2] - M[2, 0])
+        d = Quaternion.__copysign(d, M[1, 0] - M[0, 1])
+
+        return Quaternion(a, b, c, d)
+
+    @staticmethod
+    def __copysign(x, y):
+
+        # Takes the sign from the second term and sets the sign of the first
+        # without altering the magnitude.
+
+        if y < 0:
+            if x < 0:
+                return x
+            elif x > 0:
+                return -x
+            else:
+                return 0
+        elif y > 0:
+            if x > 0:
+                return x
+            elif x < 0:
+                return -x
+            else:
+                return 0
+        else:
+            return 0
+
+    def __add__(self, other):
+        return self.add(other)
+
+    def __sub__(self, other):
+        return self.add(other * -1)
+
+    def __mul__(self, other):
+        return self.mul(other)
+
+    def __pow__(self, p):
+        return self.pow(p)
+
+    def __repr__(self):
+        return sstr(self)
+
+    def _eval_Integral(self, *args):
+        return self.integrate(*args)
+
+    def _eval_diff(self, *symbols, **kwargs):
+        return self.diff(*symbols)
+
+    def add(self, other):
+        """Adds quaternions.
+
+        Example
+        ========
+
+        >>> from sympy.polys.domains.quaternion import Quaternion
+        >>> from sympy import symbols
+        >>> q1 = Quaternion(1, 2, 3, 4)
+        >>> q2 = Quaternion(5, 6, 7, 8)
+        >>> print(q1.add(q2))
+        6 + (8)i + (10)j + (12)k
+        >>> print(q1.add(5))
+        6 + (2)i + (3)j + (4)k
+        >>> x = symbols('x', real = True)
+        >>> print(q1.add(x))
+        x + 1 + (2)i + (3)j + (4)k
+        """
+        q1 = self
+        q2 = other
+
+        # If q1 or q2 is a number or a sympy expression instead of a quaternion
+        if not isinstance(q1, Quaternion):
+            return Quaternion(re(q1) + q2.a, im(q1) + q2.b, q2.c, q2.d)
+
+        if not isinstance(q2, Quaternion):
+            return Quaternion(re(q2) + q1.a, im(q2) + q1.b, q1.c, q1.d)
+
+        return Quaternion(q1.a + q2.a, q1.b + q2.b, q1.c + q2.c, q1.d
+                          + q2.d)
+
+    def mul(self, other):
+        """Multiplies quaternions.
+
+        Example
+        ========
+
+        >>> from sympy.polys.domains.quaternion import Quaternion
+        >>> from sympy import symbols
+        >>> q1 = Quaternion(1, 2, 3, 4)
+        >>> q2 = Quaternion(5, 6, 7, 8)
+        >>> print(q1.mul(q2))
+        -60 + (12)i + (30)j + (24)k
+        >>> print(q1.mul(2))
+        2 + (4)i + (6)j + (8)k
+        >>> x = symbols('x', real = True)
+        >>> print(q1.mul(x))
+        x + (2*x)i + (3*x)j + (4*x)k
+        """
+        q1 = self
+        q2 = other
+
+        # If q1 or q2 is a number or a sympy expression instead of a quaternion
+        if not isinstance(q1, Quaternion):
```

Comment:
you are assuming that `q1` is a complex, it might not be the case.

Reply:
I did this because I thought it would be better to take a general case. If real is set to true for an expression then im() will be 0.

Reply:
What if `q1` is a matrix or a Grassmann number?

Reply:
So would it be a good idea to consider the default case to be real and have a special case for complex?

Reply:
It looks like `q1 = self` would be an object of the `Quaternion` class.

Reply:
@jksuom 
Thanks for pointing that out. I should've seen that. I guess I should just check for q2 then.
One problem that I'm facing is that q*4 is working but 4*q isn't. Do u know how to fix this?

Reply:
There is a priority mechanism implemented in [core](https://github.com/sympy/sympy/blob/master/sympy/core/expr.py#L95) for that purpose. If a quaternion q has higher priority than a complex number c, then a `c*q` will be computed as `q.__rmul__(c)` instead of `c.__mul__(q)` (and it is possible to define `__rmul__ = __mul__`).

Reply:
@jksuom 
Thanks a lot! I used __radd__ and __rmul__ in my code and it worked.

Reply:
Anyway, I would add:
```python
if not q2.is_complex:
    return Add(self, q2)
```
or something similar, that is leave `quaterion + noncomplex` unevaluated.